controller/rollout_load_controller.go

```diff
@@ -0,0 +1,1579 @@
+	log := c.getLoggerForRolloutLoad()
+
+	bsRollouts, err := c.backupTargetClient.RolloutList(c.backupTargetURL, c.backupTargetCredential)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to list rollouts in the backup target")
+	}
+
+	for bsRolloutName, uri := range bsRollouts {
+		if bsRolloutName != c.rollout.Spec.Name {
+			continue
+		}
+
+		bsRolloutVersion, _, err := types.ParseRolloutURI(uri)
+		if err != nil {
+			return "", err
+		}
+
+		if bsRolloutVersion != c.rollout.Spec.Version {
+			log.Debugf("Found %v in version %v", bsRolloutName, bsRolloutVersion)
+			continue
+		}
+
+		return c.backupTargetURL + uri, nil
+	}
+
+	return "", errors.Errorf("cannot find %v/%v in %v", c.rollout.Spec.Version, c.rollout.Spec.Name, bsRollouts)
+}
+
+func (c *RolloutLoadController) restore(kind string, fn func() error, log logrus.FieldLogger, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	timeout := types.RolloutTimeout * time.Second
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	var err error
+	for {
+		select {
+		case <-timer.C:
+			c.postRestoreHandle(kind, err)
+			return
+		case <-ticker.C:
+			err = fn()
+			if err == nil {
+				c.postRestoreHandle(kind, nil)
+				return
+			}
+			log.WithError(err).Debugf(RolloutLoadMsgRestoringFmt, kind, c.rollout.Spec.Name)
+		}
+	}
+}
+
+func (c *RolloutLoadController) postRestoreHandle(kind string, restoreError error) {
+	log := c.getLoggerForRolloutLoad()
+
+	rollout, err := c.ds.GetRolloutInProgress(c.rollout.Name)
+	if err != nil {
+		restoreError = err
+	}
+
+	if rollout.Status.State == longhorn.RolloutStateError {
+		return
+	} else if restoreError == nil {
+		rolloutLoadMsgRestored := fmt.Sprintf(RolloutLoadMsgRestoredFmt, kind, c.rollout.Spec.Name)
+		log.Infof(rolloutLoadMsgRestored)
+		c.eventRecorder.Eventf(rollout, corev1.EventTypeNormal, EventReasonRestored, rolloutLoadMsgRestored)
+		return
+	}
+
+	c.handleRolloutErr(rollout, longhorn.RolloutConditionReasonRestore,
+		"failed to restore resource from rollout", restoreError, log)
+
+	_, err = c.ds.UpdateRolloutStatus(rollout)
+	if err != nil && !apierrors.IsConflict(errors.Cause(err)) {
+		log.WithError(err).Error(RolloutLoadErrorUpdateStatusFailed)
+		return
+	}
+}
+
+func (c *RolloutLoadController) restoreClusterRoles() (err error) {
+	if c.clusterRoleList == nil {
+		return nil
+	}
+
+	for _, restore := range c.clusterRoleList.Items {
+		log := c.logger.WithField(restore.Kind, restore.Name)
+
+		exist, err := c.ds.GetClusterRole(restore.Name)
+		if err != nil {
+			if !datastore.ErrorIsNotFound(err) {
+				return err
+			}
+
+			restore.ResourceVersion = ""
+
+			log.Info(RolloutLoadMsgCreating)
+			exist, err = c.ds.CreateClusterRole(&restore)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				return err
+			}
+		}
+
+		if !reflect.DeepEqual(exist.Rules, restore.Rules) {
+			log.Info(RolloutLoadMsgUpdating)
+			exist.Rules = restore.Rules
+		}
```

Comment:
We can avoid the update if there is no change against `Rules` and `Annotation`

Reply:
The idea is to do updates with or without changes from the rollout. This is to tag the rollout timestamp to the resources.

I've swapped the sequence anyways because it is better to handle timestamp taggings first if it returns error.